### PR TITLE
Maninv

### PR DESCRIPTION
**Pull request by KapilDiwakar**

### CODE REVIEW

File: core/apis/trackvia/manual_invoice.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            key_name = ref_field_mappings.get(field.get('fieldMetaId'))
+        value = field.get('value', '')
+
+        if field.get('fieldMetaId') in ref_field_mappings.keys():
+            value = field.get('identifier', '')
+
+        return_dict[key_name] = value
+
+    return_dict['tv_id'] = mi_id
+    return_dict['CONTRACTOR'] = "Diego Rodriguez"  # Final Contractor name to be decided
```
Comment (akshay25) [inline core/apis/trackvia/manual_invoice.py:62]: Need to set the right contractor name.

File: core/fieldMapper.py [hunk 2]
```diff
@@ -5,31 +5,33 @@
...
+def _invoiceMapper(invoice_data, is_manual):
+    return_dict = {
         'DocNumber': invoice_data.get('INVOICE ID'),
         'DueDate': invoice_data.get('DUE DATE'),
         'CustomerMemo': {'value': invoice_data.get('NOTES')},
         'SalesTermRef': {'value': _salesTermMapper(invoice_data.get('PAYMENT TERMS'))},
         'BillEmail': {'Address': invoice_data.get('CONTRACTOR EMAIL')},
         'TxnDate': invoice_data.get('INVOICE DATE'),
-        'TxnTaxDetail': _get_tax_details(invoice_data.get('SALE TAX'), invoice_data.get('INVOICE ID')),
+        # 'TxnTaxDetail': _get_tax_details(invoice_data.get('SALE TAX'), invoice_data.get('INVOICE ID')),
```
Comment (akshay25) [inline core/fieldMapper.py:27]: can remove this comment.

File: core/fieldMapper.py [hunk 3]
```diff
@@ -38,25 +40,49 @@ def _invoiceMapper(invoice_data):
...
     if 'Customer' in result:
-       return {'value': result['Customer']['Id']}
+        return {'value': result['Customer']['Id']}
     else:
-        logger.error('error finding customer: {0} in Quickbooks while processing trackvia invoice: {1}'.format(cust_name, tv_invoice_id))
-        send_email('TV-QBO integeration error', 'We got an error finding customer: {0} in Quickbooks while processing trackvia invoice: {1}. Invoice creation/updation failed. Please create customer in quickbooks and retry.'.format(cust_name, tv_invoice_id))
+        logger.error(
+            'error finding customer: {0} in Quickbooks while processing trackvia invoice: {1}'.format(cust_name,
+                                                                                                      tv_invoice_id))
+        send_email('TV-QBO integeration error',
```
Comment (akshay25) [inline core/fieldMapper.py:57]: no way of knowing whether it is manual or normal invoice.

File: core/tasks.py [hunk 4]
```diff
@@ -39,7 +41,10 @@ def process_tv_webhook(table_id, view_id, record_id, event_type):
                 table_id, view_id, record_id, event_type))
             return
         elif event_type == 'AFTER_UPDATE':
-            record = getFullInvoiceData(record_id, view_id)
+            if manual_invoice_view_id == view_id:
```
Comment (akshay25) [inline core/tasks.py:44]: how do we know from the callback that it is the manual or normal invoice.

File: core/tasks.py [hunk 5]
```diff
@@ -199,12 +204,22 @@ def process_invoice(invoice_id):
         return
     tv_invoice_id = invoices[0].tv_id
     view_id = invoices[0].view_id
+    is_manual = invoices[0].is_manual
     if total_amt == balance:
-        updateTvInvoiceStatus(tv_invoice_id, 'UNPAID', view_id)
```
Comment (akshay25) [inline core/tasks.py:203]: Can make two separate functions. One for normal invoice and other for manual. It will be clearer.

Comment (KapilDiwakar) [reply core/tasks.py:203]: fixed

File: core/tasks.py (hunk 4 above)
Comment (KapilDiwakar) [reply core/tasks.py:44]: fixed

File: core/fieldMapper.py (hunk 3 above)
Comment (KapilDiwakar) [reply core/fieldMapper.py:57]: fixed

File: core/fieldMapper.py (hunk 2 above)
Comment (KapilDiwakar) [reply core/fieldMapper.py:27]: fixed

File: core/apis/trackvia/manual_invoice.py (hunk 1 above)
Comment (KapilDiwakar) [reply core/apis/trackvia/manual_invoice.py:62]: fixed. removed this line

**Pull request merged**